File: worker/lib/bus.ts
import type Redis from "ioredis";

/** Realtime message published on `rt:{uid}` (→ SSE) and appended to `events:{uid}` stream. */
export interface RtMessage {
  jobId: string;
  type: "progress" | "done" | "error";
  step?: string;
  status?: string;
  result?: unknown;
  error?: string;
  at: number;
}

export interface Bus {
  publish: (uid: string, msg: Omit<RtMessage, "at">) => Promise<void>;
}

/** Pub/sub + event-stream bridge to the UI. */
export function makeBus(redis: Redis): Bus {
  return {
    async publish(uid, msg) {
      const full: RtMessage = { ...msg, at: Date.now() };
      const payload = JSON.stringify(full);
      try {
        await redis.publish(`rt:${uid}`, payload);
        await redis.xadd(`events:${uid}`, "MAXLEN", "~", 500, "*", "data", payload);
      } catch (e) {
        console.warn("[bus] publish failed:", (e as Error).message);
      }
    },
  };
}

File: worker_py/bus.py
"""Realtime bus → UI. Port of ``worker/lib/bus.ts``.

Publishes each message on the ``rt:{uid}`` pub/sub channel (bridged to the
browser by ``/api/events`` SSE) and appends it to the capped ``events:{uid}``
stream. The message shape must match ``RtMessage`` so the client parses it.
"""

import json
from typing import Any, Dict

from redis.asyncio import Redis

from .util import now_ms


async def publish(redis: Redis, uid: str, msg: Dict[str, Any]) -> None:
    """Publish a realtime message (``at`` is stamped here, like the Node bus)."""
    full = {**msg, "at": now_ms()}
    payload = json.dumps(full)
    try:
        await redis.publish(f"rt:{uid}", payload)
        await redis.xadd(f"events:{uid}", {"data": payload}, maxlen=500, approximate=True)
    except Exception as e:  # noqa: BLE001
        print(f"[bus] publish failed: {e}")
